livedata-ktx/src/main/java/com/shopify/livedataktx/SupportMediatorLiveData.kt

```diff
@@ -28,28 +28,73 @@ import android.arch.lifecycle.LifecycleOwner
 import android.arch.lifecycle.MediatorLiveData
 import android.arch.lifecycle.Observer
 
-open class SupportMediatorLiveData<T>(internal val isSingle: Boolean = false, private val versionProvider: (() -> Int)? = null) : MediatorLiveData<T>() {
+private typealias ObserverWrappers<T> = MutableList<Pair<Observer<T>, Observer<T>>>
+
+private fun <T> ObserverWrappers<T>.get(observer: Observer<T>): Observer<T>? =
+    indexOfFirst { it.first == observer }
+        .takeUnless { it < 0 }
+        ?.let { this[it].second }
+
+private fun <T> ObserverWrappers<T>.getOrElse(
+    observer: Observer<T>,
+    wrapper: Observer<T>
+): Observer<T> = get(observer) ?: wrapper
+
+private fun <T> ObserverWrappers<T>.putIfAbsent(observer: Observer<T>, wrapper: Observer<T>) {
+    val index = indexOfFirst { it.first == observer }
+    if (index < 0) {
+        add(Pair(observer, wrapper))
+    }
+}
+
+private fun <T> ObserverWrappers<T>.removeIfPresent(observer: Observer<T>) {
+    val index = indexOfFirst { it.first == observer }
+    if (index >= 0) {
+        removeAt(index)
+    }
+}
+
+open class SupportMediatorLiveData<T>(
+    internal val isSingle: Boolean = false,
+    private val versionProvider: (() -> Int)? = null
+) : MediatorLiveData<T>() {
+
+    private val observerWrappers: ObserverWrappers<T> = mutableListOf()
 
     private var _version = 0
     internal val version: Int get() = versionProvider?.let { it() } ?: _version
 
     @Deprecated("Use observe extension")
     override fun observe(owner: LifecycleOwner, observer: Observer<T>) {
         val observerVersion = version
-        super.observe(owner, Observer {
+        val wrapper = observerWrappers.getOrElse(observer, Observer {
             if (!isSingle || observerVersion < version) {
                 observer.onChanged(it)
             }
         })
+        observerWrappers.putIfAbsent(observer, wrapper)
+        super.observe(owner, wrapper)
     }
 
     @Deprecated("Use observe extension without LifecycleOwner")
     override fun observeForever(observer: Observer<T>) {
-        super.observeForever(observer)
+        val observerVersion = version
+        val wrapper = observerWrappers.getOrElse(observer, Observer {
```

Comment:
nit: this snippet is duplicated from above, maybe we can make it into a function?